File: apps/ios/AnpusFeather/Game/Pyramid.swift
import CoreGraphics

/*
  Background imagery — the port of pyramid.pde. The two backdrop photos scale
  to frame the sides at any width; Anubis blinks between two frames in the
  center.
*/
final class Pyramid {
    private var tintAlpha: CGFloat = 0
    private var clock: CGFloat = 0
    private let store: TextureStore

    init(store: TextureStore) { self.store = store }

    func tick(_ dt: CGFloat) { clock += dt }
    func setAlpha(_ a: CGFloat) { tintAlpha = a }

    private var tint: Color4 { Color4(r: 1, g: 1, b: 1, a: Float(tintAlpha / 255)) }

    func show(_ canvas: Canvas) {
        guard let p0 = store.image("pyramid0"), let p2 = store.image("pyramid2") else { return }
        let half = (World.w / 2).rounded(.up) + 1
        let lh = CGFloat(p2.height) * (half / CGFloat(p2.width))
        let rh = CGFloat(p0.height) * (half / CGFloat(p0.width))
        canvas.drawImage(p2, 0, 0, half, lh, tint: tint)
        canvas.drawImage(p0, World.w - half, 0, half, rh, tint: tint)

        let name = clock.truncatingRemainder(dividingBy: 1.1) < 0.55 ? "anubis1" : "anubis2"
        if let a = store.image(name) {
            let w = CGFloat(a.width)
            let h = CGFloat(a.height)
            canvas.drawImage(a, World.w / 2 - w / 2, World.h / 2 - 100 - h / 2, w, h, tint: tint)
        }
    }

    func showAlt(_ canvas: Canvas) {
        guard let img = store.image("pyramid3") else { return }
        let scale = World.w / CGFloat(img.width)
        let h = CGFloat(img.height) * scale
        canvas.drawImage(img, 0, World.h - h, World.w, h, tint: tint)
    }

    func fadeIn(_ dt: CGFloat) { tintAlpha = clampf(tintAlpha + 5 * dt * FPS, 0, 255) }
    func fadeOut(_ dt: CGFloat) { tintAlpha = clampf(tintAlpha - 10 * dt * FPS, 0, 255) }
}

File: apps/unity/Assets/AnpusFeather/Scripts/Pyramid.cs
using UnityEngine;

namespace AnpusFeather
{
    /*
      Background imagery — port of pyramid.pde.
    */
    public sealed class Pyramid
    {
        float _tintAlpha;
        float _clock;
        readonly AssetStore _store;

        public Pyramid(AssetStore store) => _store = store;

        public void Tick(float dt) => _clock += dt;
        public void SetAlpha(float a) => _tintAlpha = a;

        Color4 Tint => new Color4(1f, 1f, 1f, _tintAlpha / 255f);

        public void Show(DrawCanvas canvas)
        {
            Texture2D p0 = _store.Image("pyramid0");
            Texture2D p2 = _store.Image("pyramid2");
            if (p0 == null || p2 == null) return;

            float half = Mathf.Ceil(World.W / 2f) + 1f;
            float lh = p2.height * (half / p2.width);
            float rh = p0.height * (half / p0.width);
            canvas.DrawImage(p2, 0f, 0f, half, lh, Tint);
            canvas.DrawImage(p0, World.W - half, 0f, half, rh, Tint);

            string name = _clock % 1.1f < 0.55f ? "anubis1" : "anubis2";
            Texture2D anubis = _store.Image(name);
            if (anubis != null)
            {
                float w = anubis.width;
                float h = anubis.height;
                canvas.DrawImage(anubis, World.W / 2f - w / 2f, World.H / 2f - 100f - h / 2f, w, h, Tint);
            }
        }

        public void ShowAlt(DrawCanvas canvas)
        {
            Texture2D img = _store.Image("pyramid3");
            if (img == null) return;
            float scale = World.W / img.width;
            float h = img.height * scale;
            canvas.DrawImage(img, 0f, World.H - h, World.W, h, Tint);
        }

        public void FadeIn(float dt) => _tintAlpha = MathUtil.Clamp(_tintAlpha + 5f * dt * World.Fps, 0f, 255f);
        public void FadeOut(float dt) => _tintAlpha = MathUtil.Clamp(_tintAlpha - 10f * dt * World.Fps, 0f, 255f);
    }
}
